designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,75 @@
+- Start Date: 2019-09-28
+- RFC PR: https://github.com/eslint/rfcs/pull/40
+- Authors: Toru Nagashima ([@mysticatea](https://github.com/mysticatea))
+
+# Moving to Asynchronous API
+
+## Summary
+
+This RFC adds a new class `ESLint` that has asynchronous API and deprecates `CLIEngine`.
+
+## Motivation
+
+- Dynamic `import()` has arrived at Stage 4. The dynamic loading of ES modules requires asynchronous API. Migrating to asynchronous API opens up doors to write plugins/configs with ES modules.
+- Linting in parallel requires asynchronous API. We can improve linting logic to run it in parallel. And we can improve config loading logic and file enumeration logic to run it in parallel. (E.g., load `extends` list, traverse child directories.)
+
+Because the migration of public API needs long time, we should start to migrate our APIs earlier.
+
+And the name of `CLIEngine`, our primary API, causes confusion to the community. People try `Linter` class at first, then they notice it doesn't work as expected. We have a lot of issues that say "please use `CLIEngine` instead."
+The name of new class, `ESLint`, is our primary API clearly.
+
+## Detailed Design
+
+### Add new `ESLint` class
+
+This RFC adds a new class `ESLint`. It has almost the same methods as `CLIEngine`, but the following methods return `Promise`.
+
+- `executeOnFiles()`
+- `executeOnText()`
+- `getConfigForFile()`
+- `getFormatter()`
+- `isPathIgnored()`
+- `outputFixes()`
+
+The following methods are as-is because those don't touch both file system and module system.
+
+- `addPlugin()`
+- `getErrorResults()`
```

Comment:
One question I had about introducing async APIs is if we should also consider streaming output as the same time. But I guess we can do it as a separate RFC.

Reply:
Ah, good catch. If we want to support streaming in our formatters and custom formatters, we should update the type of `getFormatter()` method. This is good timing for that.

Reply:
How would we be able to support both old format that would be returning an object, and new one that returns stream at the same time? What will happen if you load old formatter using new API?

Reply:
Wrapper adapts the interface (https://github.com/eslint/rfcs/pull/40/files#diff-a206adca92f13d7d6f95fe808b733c13R40-R51).

Once the formatter got streaming support, we can update the wrapper implementation without any breakage.

Reply:
Sorry, to ask a more specific question. How can this API support both our internal formatters that might be converted to return streams, and third-party formatter that will return an object at the same time?

Reply:
Sorry, what is "third-party formatter that will return an object"? Formatters return a string.

Reply:
Oh, sorry, my bad, I mixed up formatter with linting result output. But question is still valid, I think. We will need to support both formatters that return a string as well as a stream at the same time (even if we convert our own formatters to streams, custom formatters still might return a string).

Reply:
Checking the returned value is very easy.

<details><summary>For example:</summary>

```js
  async getFormatter(name) {
    const format = this.cliEngine.getFormatter(name)

    // Return the wrapper.
    return async function* formatter(resultIterator) {
      // Collect the results.
      const results = []
      for await (const result of resultIterator) {
        results.push(result)
      }
      results.sort(byFilePath)

      // Make `rulesMeta`.
      const rules = this.cliEngine.getRules()
      const rulesMeta = getRulesMeta(rules)

      // Format the results with the formatter of the current spec.
      const ret = format(results, { rulesMeta })
      if (typeof ret === "string") {
        yield ret
      } else {
        yield* ret
      }
    }
  }
```

</details>

On the other hand, we need a trick to distinguish the formatter supports the iterator argument. In my head, we can do it with a new formatter style that is an object that has the `format` method.

<details><summary>For example:</summary>

**New style formatter:** (has `version` property and `format` method)

```js
module.exports = {
  version: 2,
  async *format(resultsIterator, metadata) {
    const results = []
    let totalCount = 0
    let errorCount = 0

    // Show progress state.
    yield progress(totalCount, errorCount)
    for await (const result of resultsIterator) {
      results.push(result)
      totalCount += 1
      totalCount += result.errorCount
      yield progress(totalCount, errorCount)
    }

    // Show results.
    results.sort(byFileName)
    yield formatResults(results)
  }
}
```

**`getFormatter()` function that handles the new style formatter:**

```js
  async getFormatter(name) {
    const formatter = this.cliEngine.getFormatter(name)

    if (typeof formatter === "function") {
      // Return the adapter for legacy style formatter.
      return async function* adapter1(resultIterator) {
        // Collect the results.
        const results = []
        for await (const result of resultIterator) {
          results.push(result)
        }
        results.sort(byFilePath)

        // Make `rulesMeta`.
        const rules = this.cliEngine.getRules()
        const rulesMeta = getRulesMeta(rules)

        // Format the results with the formatter of the current spec.
        yield formatter(results, { rulesMeta })
      }
    }

    if (formatter.version === 2) {
      // Return the adapter for new style formatter.
      return async function* adapter2(results) {
        const rules = this.cliEngine.getRules()
        const rulesMeta = getRulesMeta(rules)
        const ret = formatter.format(results, { rulesMeta })
        if (typeof ret === "string") {
          yield ret
        } else {
          yield* ret
        }
      }
    }

    throw new Error("unknown version number: " + formatter.version)
  }
```

</details>